cylc/flow/ws_data_mgr.py

```diff
@@ -0,0 +1,393 @@
+from cylc.flow.task_job_logs import JOB_LOG_OPTS
+from cylc.flow import __version__ as CYLC_VERSION
+from cylc.flow.ws_messages_pb2 import (
+    PbFamily, PbFamilyProxy, PbTask, PbTaskProxy, PbWorkflow,
+    PbEdge, PbEdges, PbEntireWorkflow)
+
+
+class WsDataMgr(object):
+    """Manage the data feed for the User Interface Server."""
+
+    TIME_FIELDS = ['submitted_time', 'started_time', 'finished_time']
+
+    def __init__(self, schd):
+        self.schd = schd
+        self.tasks = {}
+        self.task_proxies = {}
+        self.task_states = {}
+        self.families = {}
+        self.family_proxies = {}
+        self.state_count_totals = {}
+        self.state_count_cycles = {}
+        self.edges = {}
+        self.graph = PbEdges()
+        self.workflow_id = f"{self.schd.owner}/{self.schd.suite}"
+        self.workflow = PbWorkflow()
+
+    def initiate_data_model(self):
+        """Initiate or Update data model."""
+        update_time = time()
+        config = self.schd.config
+        tasks = {}
+        task_proxies = {}
+        families = {}
+        family_proxies = {}
+        edges = {}
+        graph = PbEdges()
+        workflow = PbWorkflow(
+            checksum=f"{self.workflow_id}@{update_time}",
+            id=self.workflow_id,
+        )
+
+        all_states = []
+
+        # Compute state_counts (total, and per cycle).
+        state_count_totals = {}
+        state_count_cycles = {}
+
+        ancestors_dict = config.get_first_parent_ancestors()
+        descendants_dict = config.get_first_parent_descendants()
+        parents_dict = config.get_parent_lists()
+
+        # create task definition data objects
+        for name, tdef in config.taskdefs.items():
+            t_id = f"{self.workflow_id}/{name}"
+            t_check = f"{name}@{update_time}"
+            task = PbTask(
+                checksum=t_check,
+                id=t_id,
+                name=name,
+                depth=len(ancestors_dict[name]) - 1,
+            )
+            task.namespace.extend(copy(tdef.namespace_hierarchy))
+            for key, val in dict(tdef.describe()).items():
+                if key in ['title', 'description', 'URL']:
+                    setattr(task.meta, key, val)
+                else:
+                    task.meta.user_defined.append(f"{key}={val}")
+            ntimes = len(tdef.elapsed_times)
+            if ntimes:
+                task.mean_elapsed_time = (
+                    float(sum(tdef.elapsed_times)) / ntimes)
+            elif tdef.rtconfig['job']['execution time limit']:
+                task.mean_elapsed_time = \
+                    tdef.rtconfig['job']['execution time limit']
+            tasks[name] = task
+
+        # create task definition data objects
+        for itask in self.schd.pool.get_all_tasks():
+            ts = itask.get_state_summary()
+            name, point_string = TaskID.split(itask.identity)
+            if name not in tasks:
+                continue
```

Comment:
Does it deserve a log entry, maybe in debug level, or is that OK if the schedule pool has tasks that were not found in the configuration?

Reply:
Removed equivalent. Should be safe without